subsys/bluetooth/mesh/adv_ext.c

```diff
@@ -69,72 +69,71 @@ struct bt_mesh_ext_adv {
 
-static STRUCT_SECTION_ITERABLE(bt_mesh_ext_adv, adv_main) = {
-	.tag = (
+static struct bt_mesh_ext_adv advs[] = {
+	[0] = {
+		.tags = (
 #if !defined(CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE)
-		BT_MESH_FRIEND_ADV |
+			BT_MESH_ADV_TAG_BIT_FRIEND |
 #endif
 #if !defined(CONFIG_BT_MESH_ADV_EXT_GATT_SEPARATE)
-		BT_MESH_PROXY_ADV |
+			BT_MESH_ADV_TAG_BIT_PROXY |
 #endif /* !CONFIG_BT_MESH_ADV_EXT_GATT_SEPARATE */
 #if defined(CONFIG_BT_MESH_ADV_EXT_RELAY_USING_MAIN_ADV_SET)
-		BT_MESH_RELAY_ADV |
+			BT_MESH_ADV_TAG_BIT_RELAY |
 #endif /* CONFIG_BT_MESH_ADV_EXT_RELAY_USING_MAIN_ADV_SET */
-		BT_MESH_LOCAL_ADV),
-
-	.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
-};
-
+#if defined(CONFIG_BT_MESH_PB_ADV)
+			BT_MESH_ADV_TAG_BIT_PROV |
+#endif /* CONFIG_BT_MESH_PB_ADV */
+			BT_MESH_ADV_TAG_BIT_LOCAL
+		),
+		.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
+	},
 #if CONFIG_BT_MESH_RELAY_ADV_SETS
-static STRUCT_SECTION_ITERABLE_ARRAY(bt_mesh_ext_adv, adv_relay, CONFIG_BT_MESH_RELAY_ADV_SETS) = {
-	[0 ... CONFIG_BT_MESH_RELAY_ADV_SETS - 1] = {
-		.tag = BT_MESH_RELAY_ADV,
+	[1 ... CONFIG_BT_MESH_RELAY_ADV_SETS] = {
+		.tags = (
+#if defined(CONFIG_BT_MESH_RELAY)
+			BT_MESH_ADV_TAG_BIT_RELAY |
+#endif /* CONFIG_BT_MESH_RELAY */
+#if defined(CONFIG_BT_MESH_PB_ADV_USE_RELAY_SETS)
+			BT_MESH_ADV_TAG_BIT_PROV |
+#endif /* CONFIG_BT_MESH_PB_ADV_USE_RELAY_SETS */
+		0),
 		.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
-	}
-};
+	},
 #endif /* CONFIG_BT_MESH_RELAY_ADV_SETS */
-
 #if defined(CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE)
-#define ADV_EXT_FRIEND 1
-static STRUCT_SECTION_ITERABLE(bt_mesh_ext_adv, adv_friend) = {
-	.tag = BT_MESH_FRIEND_ADV,
-	.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
-};
-#else /* CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE */
-#define ADV_EXT_FRIEND 0
+	{
+		.tags = BT_MESH_ADV_TAG_BIT_FRIEND,
+		.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
+	},
 #endif /* CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE */
-
 #if defined(CONFIG_BT_MESH_ADV_EXT_GATT_SEPARATE)
-#define ADV_EXT_GATT 1
-static STRUCT_SECTION_ITERABLE(bt_mesh_ext_adv, adv_gatt) = {
-	.tag = BT_MESH_PROXY_ADV,
-	.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
-};
-#else /* CONFIG_BT_MESH_ADV_EXT_GATT_SEPARATE */
-#define ADV_EXT_GATT 0
+	{
+		.tags = BT_MESH_ADV_TAG_BIT_PROXY,
+		.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
+	},
 #endif /* CONFIG_BT_MESH_ADV_EXT_GATT_SEPARATE */
+};
 
-#define BT_MESH_ADV_COUNT (1 + CONFIG_BT_MESH_RELAY_ADV_SETS + ADV_EXT_FRIEND + ADV_EXT_GATT)
-
-BUILD_ASSERT(CONFIG_BT_EXT_ADV_MAX_ADV_SET >= BT_MESH_ADV_COUNT,
+BUILD_ASSERT(ARRAY_SIZE(advs) <= CONFIG_BT_EXT_ADV_MAX_ADV_SET,
 	     "Insufficient adv instances");
 
 static inline struct bt_mesh_ext_adv *relay_adv_get(void)
 {
-#if CONFIG_BT_MESH_RELAY_ADV_SETS
-	return adv_relay;
-#else /* !CONFIG_BT_MESH_RELAY_ADV_SETS */
-	return &adv_main;
-#endif /* CONFIG_BT_MESH_RELAY_ADV_SETS */
+	if (!!(CONFIG_BT_MESH_RELAY_ADV_SETS)) {
```

Comment:
What if `CONFIG_BT_MESH_RELAY_ADV_SETS` is not defined? It can happen if BT_MESH_RELAY is not enabled:
https://github.com/zephyrproject-rtos/zephyr/blob/c938a53e1d0ce03df027351aa1c487c405d6b36c/subsys/bluetooth/mesh/Kconfig#L106-L110

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/adv_ext.c#L32

Reply:
No problem for this.